extension/api.js

```diff
@@ -78,149 +105,215 @@ const lostEngines = new Map ([
+      a.installDate < lostEngines.get(a.id)
+    );
+  if (!addons.length) {
+    console.log("reset-default-search: No addons in our list are installed.");
+    finish();
+    return;
+  }
 
-    // Filter out any that are blocklisted.
-    let enabledAddons = addons.filter(a => !a.appDisabled);
+  // Make sure the setting is not user controlled.
+  await ExtensionSettingsStore.initialize();
+  let control = await ExtensionSettingsStore.getLevelOfControl(
+    ExtensionSettingsStore.SETTING_USER_SET,
+    DEFAULT_SEARCH_STORE_TYPE,
+    DEFAULT_SEARCH_SETTING_NAME
+  );
+  if (control == "controlled_by_this_extension") {
+    console.log(`reset-default-search: search control has been user selected, not resetting, finished.`);
+    finish();
+    return;
+  }
 
-    // If the extension is not running or enabled in the search service
-    // we will try again later.
-    let tryagain = false;
+  // Filter out any that are blocklisted.
+  let enabledAddons = addons.filter(a => !a.appDisabled);
+  console.log(`reset-default-search: enabled and eligible addons ${enabledAddons.length}`);
 
-    // We will only ask for the latest installed engine.  We will
-    // loop through the list until one of them makes it to the prompt.
-    enabledAddons.sort((a, b) => b.installDate - a.installDate);
-    for (let addon of enabledAddons) {
-      console.log(`reset-default-search: reset search engine to ${addon.id}`);
+  // We will only ask for the latest installed engine.  We will
+  // loop through the list until one of them makes it to the prompt.
+  enabledAddons.sort((a, b) => b.installDate - a.installDate);
+  for (let addon of enabledAddons) {
+    console.log(`reset-default-search: reset search engine to ${addon.id}`);
 
-      let policy = WebExtensionPolicy.getByID(addon.id);
-      if (!policy?.extension) {
-        tryagain = true;
-        console.log("reset-default-search: extension is not running, cannot set as default");
-        continue;
-      }
-      let { extension } = policy;
-      let { manifest } = extension;
+    let policy = WebExtensionPolicy.getByID(addon.id);
+    if (!policy?.extension) {
+      tryagain = true;
+      console.log("reset-default-search: extension is not running, cannot set as default, try again later");
+      continue;
+    }
+    let { extension } = policy;
+    let { manifest } = extension;
 
-      let searchProvider = manifest?.chrome_settings_overrides?.search_provider;
-      if (!searchProvider?.is_default) {
-        // If the extension isn't asking to be default at this point, bail out.
-        console.log("reset-default-search: is_default is not requested by the addon");
-        continue;
-      }
+    let searchProvider = manifest?.chrome_settings_overrides?.search_provider;
+    if (!searchProvider?.is_default) {
+      // If the extension isn't asking to be default at this point, bail out.
+      console.log("reset-default-search: is_default is not requested by the addon, try again later");
```

Comment:
Do we need to set `tryagain = true` here?

Reply:
no, this instance wouldn't be eligible.  it will continue to the next if there are more, and on exit it will set tryagain if there are still blocklisted addons.